### ISSUE TITLE
[FEA] Enable CUDA UVM (Unified/Universal) Memory Support from Numba

### ISSUE THREAD

**Issue by awthomp** [numba/numba#4362]
## Feature request
NVIDIA's embedded GPU line (TX2, Xavier, Nano, to name a few) feature a shared memory space between CPU and GPU. Typically handled in CUDA with unified memory, data access between host and device involves a zero-copy.

On these embedded platforms (I'm using the TX2), I'm seeing different pointer addresses when analyzing the pointer from \_\_array_interface\_\_ and \_\_cuda_array_interface\_\_.

```
>>> import numpy as np
>>> import cupy as cp
>>> from numba import cuda
>>> 
>>> arr = np.arange(100)
>>>
>>> arr.__array_interface__['data']
(94169420899168, False)
>>> 
>>> grr = cuda.to_device(arr)
>>> 
>>> grr.__cuda_array_interface__['data']
(140351520112640, False)
```

`numba.cuda.to_device()` doesn't seem to have awareness that the memory allocated via `np.arange()`, in this case, is also accessible to the GPU. 

Instead, we see a performance overhead due to a physical memory copy and different pointer address.

## Test Setup
NVIDIA Jetson TX2
Numba 0.45 via [conda4aarch64](https://github.com/jjhelmus/conda4aarch64/releases)

**Comment by seibert** [comment]
Seems like we could look at the device properties to determine whether a copy is needed.  This will also affect the CUDA Python kernels, as Numba automatically calls to_device() when host arrays are passed to the GPU for processing.

**Comment by seibert** [comment]
It would be nice if the property we use to detect this situation also worked when the future support for automatic paging of host memory to the device appears in Linux x86_64.

**Comment by seibert** [comment]
Also, do you know if zero-copy works on Linux POWER8/9 when the GPU is connected to the CPU over NVLink?  There may be multiple platforms affected by this, now that I think about it.

**Comment by awthomp** [comment]
I don't *think* POWER CPU/GPU is a shared memory space via NVLINK, but don't quote me on that one. What you do get is massively increased BW over a PCIe Gen3 data movement. Each NVLINK lane is 50GB/s bi-directional.

Happy to see some great engagement on this feature request, @seibert!

**Comment by stuartarchibald** [comment]
CC @andih, you may be interested.

**Comment by seibert** [comment]
Did some hunting, and I think I was remembering this talk from GTC:

http://on-demand.gputechconf.com/gtc/2018/presentation/s8430-everything-you-need-to-know-about-unified-memory.pdf

Specifically slide 47.  However the requirements are pretty strict: POWER9, Volta, and NVLINK2.  Also not sure if the Address Translation Service is enabled by default.

**Comment by awthomp** [comment]
Looping in @kkraus14 

I can do some digging too, but unified memory is certainly also supported on Pascal architectures. This is just a reminder that I don't know nearly as much about GPU memory hierarchy as I thought :).

**Comment by seibert** [comment]
OK, looking at these docs:

https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1gb22e8256592b836df9a9cc36c9db7151

I think the attribute we need to check is cudaDevAttrUnifiedAddressing?  Not sure if cudaDevAttrConcurrentManagedAccess is also relevant.  There a lot of device attributes since the last time I looked.  😅

**Comment by seibert** [comment]
actually, I think we want cudaDevAttrIntegrated, at least for the Tegra chips.

**Comment by awthomp** [comment]
I sent out the call for help, Stan. I'm over my skis on this one!

**Comment by awthomp** [comment]
Okay, did some asking around and got the following information:

It's probably desirable (at some level) to have three different types of GPU memory allocators:
1. normal cudaMalloc
2. zeroCopy (cudaHostAlloc)
3. cudaMallocManaged

You could have these different allocations be either user-defined or check the device for which allocator to use.

For Tegra, the 'easiest' is cudaManaged (Managed Memory, aka Unified Memory) as the Tegra driver is smart enough not to move data.

**Comment by awthomp** [comment]
@mnicely found a nice work-around which may give some ideas for general handling of this feature request:

```
from numba import cuda
import numpy as np
import cupy as cp

# Number of elements in array
N = 10

arr = np.arange(N)
gpu_arr = cuda.mapped_array(N, dtype=np.float, strides=None, order='C', stream=0, portable=False, wc= True)

# copy contents of arr to a
gpu_arr[0:10] = arr

# print gpu_arr
gpu_arr
>>> MappedNDArray([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])

# look at a pointers
gpu_arr.__array_interface__['data']
>>> (140323521036288, False)

gpu_arr.__cuda_array_interface__['data']
>>> (140323521036288, False)

# cupy zero copy works too
cupy_arr = cp.asarray(gpu_arr)
cupy_arr.__cuda_array_interface__['data']
>>> (140323521036288, False)
```

**Comment by seibert** [comment]
I think these suggestions are going at the problem in the other direction from where we started.

Originally, we were trying to avoid an unnecessary copy of a NumPy array when processing it on a Tegra GPU.  NumPy always uses the system allocator to allocate memory, so there isn't a way to have NumPy use one of the CUDA managed memory allocators in the first place.  If Numba can determine when it is OK to pass a pointer allocated by the OS directly to the GPU, then we should be able to treat NumPy arrays as if they were device arrays and avoid any copies.

**Comment by awthomp** [comment]
@seibert -- my point was if the memory is originally allocated with something like cuda.mapped_array(), then you're already at zero copy. I don't know if this is actual doable with the numba internals and how it interacts with numpy arrays, however.

For more context, I'm essentially trying to take cupy FFTs off of ~16384 complex samples.

```
# allocate memory
sig_array = cuda.mapped_array(16384, dtype=np.complex128, strides=None, order='C', stream=0, portable=False, wc=True)

# bring data into shared-mem
sig = np.random.rand(16384) + 1j*np.random.rand(16384)
sig_array[0:16384] = sig

# Operate numpy fft
numpy_fft = np.fft.fft(sig_array)

# Operate cupy fft
cupy_array = cp.asarray(sig_array)
cupy_fft = cp.fft.fft(cupy_array)
```
Ultimately, in my example here, we should probably be able to get rid of this: `sig_array[0:16384] = sig`.

**Comment by AndiH** [comment]
Late to the party, but thanks, @stuartarchibald for the mention.

I'd consider the Tegra memory behaviour a special case (albeit the most convenient one of all 😃). _Usual_ Unified Memory does involve copies, but they are handled by the driver on-demand (via page faults).

That is also the case for POWER8, where NVLink does exist between CPU and GPU, but it's comparable _dumb_. NVLink2 on POWER9 CPU↔GPU does support ATS, which omits the page-faulting mechanism. I'm not sure how the exact behaviour with ATS is, as I don't have proper access to a POWER9 machine right now; but the memories are physically distinct so some means of copy does need to happen. @LenaO* does have access, and she's doing quite a lot of experiments with Numba on the machine currently – she might be willing to test-run and help out with things.

That said, doesn't it all boil down to enabling Unified Memory in Numba such that in the case of Tegra, Numba can use this automatically (with the CUDA API methods mentioned above), and have it in place for (future) smarter CPU-GPU memory management (ATS, [HMM](https://www.kernel.org/doc/html/latest/vm/hmm.html))? Added benefit: Have it available for people wanting to use it over usual PCIe busses (oversubscription, …).

_*: I hope that's actually her. Will update if not._

**Comment by awthomp** [comment]
> That said, doesn't it all boil down to enabling Unified Memory in Numba such that in the case of Tegra, Numba can use this automatically (with the CUDA API methods mentioned above), and have it in place for (future) smarter CPU-GPU memory management (ATS, HMM)?

I agree with this. Providing the option for Unified Memory access within Numba and specifically defaulting to that option if the user is on a Tegra + Arm SoC (so, in this case, the driver knows not to migrate memory via page-faults) would be great!

**Comment by LenaO** [comment]
Thanks to @AndiH  for mention. 

From my experience, also on Power9 unified memory *most of the time* involves a memory transfer - however, the difference to other systems is  that is not necessarily required. 
You can use "hints" to make a more precise suggested, were the memory should be located  - in this case, page faults can be avoided.   On a GPU - on Power9 Systems - the CPU can also access it without a copy (the main difference) - but I have to redo some experiments. 
Talking form a C perspective, the GPU can now also access memory allocated with malloc.


Btw. I have an early stage Bachelor-Thesis working on integrating unified memory into Numba - and we want to evaluate it on both kinds of machines (with and without Hardware Support for unified memory) - probably a Jetson would be an interesting extension...  
(And I am actually willing to  help)

**Comment by seibert** [comment]
Just to follow up: Can someone confirm that `cudaDevAttrIntegrated` is the correct device attribute to look at to determine if the device functions like a Tegra?

**Comment by awthomp** [comment]
@seibert -- yep, `cudaDevAttrIntegrated` is the correct device attribute (confirmed by folks on the Jetson team).

**Comment by awthomp** [comment]
I'm going to do a bit of cleanup here.

The mapped_array trick I was using should theoretically not migrate data from CPU->GPU on a Jetson. There will be some PCIe activity on a dedicated GPU (e.g. V100) - but you've got mapped and pinned memory, so it's really just DMA traffic. Fast!

I think the best way to address this feature request is to enable -- and we've discussed this -- unified memory support. On the Jetsons, the driver should be smart enough not to double-copy memory. As far as I know, the mapped_array functionality takes memory pages away from the OS, which could be dangerous on the Jetson.

Plus, the community may find unified memory valuable when creating custom CUDA kernels with Numba.

**Comment by awthomp** [comment]
Hi everyone -- I'm updating this feature request since first writing. My original title,

> NVIDIA Jetson SoC (TX2, Xavier, Nano) Failure to Zero-Copy from CPU to GPU with numba.cuda.to_device()

Is more aptly described with enabling NVIDIA UVM (Unified/Universal memory) within Numba. The basics of UVM can be found [here](https://devblogs.nvidia.com/unified-memory-cuda-beginners/), but I'm interested in enabling support as the NVIDIA Jetson GPU line share GPU/CPU memory space. Enabling UVM would allow for zero-copy between the GPU and CPU as the driver "knows" the memory space can be addressed by both processors.

My current work-around in [cuSignal](https://github.com/rapidsai/cusignal) is to use cuda.mapped_array, but this removes memory pages from the CPU and probably isn't the best long term solution.

**Comment by gmarkall** [comment]
Closing this as it is now implemented via #5900.
